### PR TITLE
Wrong store ID used in Order Address Renderer

### LINKED ISSUE #38412: [Issue] Wrong store ID used in Order Address Renderer
This issue is automatically created based on existing pull request: magento/magento2#37932: Wrong store ID used in Order Address Renderer

---------
### Description (*)

`\Magento\Sales\Model\Order\Address\Renderer::format` allows to format order address. Internally, the method retrieve the order store ID and update the current store ID. This produces strange behavior if used in a context with a different store as the current store is not set to its previous value at the end of the execution of the method.

The class `\Magento\Customer\Model\Address\Config` is used within the format method in order to resolve the address format to use accordingly to the order store. This class allows to set the store context thanks to `\Magento\Customer\Model\Address\Config::setStore`. However this method was not called. Here we call it and reverse the value to its original state at the end of the execution, in order to prevent weird behavior.
With this solution, the current store in the manager is not altered and the global context is preserved.
Previously, when the method `\Magento\Customer\Model\Address\Config::getFormatByCode` was called, by the method format, the store was resolved at the first call : 

```
-> \Magento\Customer\Model\Address\Config::getFormatByCode
--> \Magento\Customer\Model\Address\Config::getFormats
---> \Magento\Customer\Model\Address\Config::getStore
```

```php
public function getStore()
{
    if ($this->_store === null) {
        $this->_store = $this->_storeManager->getStore();
    }
    return $this->_store;
}
```

I guess that's why there was a `$this->storeManager->setCurrentStore($orderStore);` in `\Magento\Sales\Model\Order\Address\Renderer::format`.  

This MR aims to remove the usage of $this->storeManager->setStore which may produce unwanted side-effects and use proper method available in the `\Magento\Customer\Model\Address\Config` API.

### Fixed Issues (if relevant)

Cases where multiple order mails (with different stores) have to be sent by the cron job.
Order addresses are not formatted accordingly to their respective store ID.
Different formats must be set by store ID.

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by thomas-kl1**
### Description (*)

`\Magento\Sales\Model\Order\Address\Renderer::format` allows to format order address. Internally, the method retrieve the order store ID and update the current store ID. This produces strange behavior if used in a context with a different store as the current store is not set to its previous value at the end of the execution of the method.

The class `\Magento\Customer\Model\Address\Config` is used within the format method in order to resolve the address format to use accordingly to the order store. This class allows to set the store context thanks to `\Magento\Customer\Model\Address\Config::setStore`. However this method was not called. Here we call it and reverse the value to its original state at the end of the execution, in order to prevent weird behavior.
With this solution, the current store in the manager is not altered and the global context is preserved.
Previously, when the method `\Magento\Customer\Model\Address\Config::getFormatByCode` was called, by the method format, the store was resolved at the first call : 

```
-> \Magento\Customer\Model\Address\Config::getFormatByCode
--> \Magento\Customer\Model\Address\Config::getFormats
---> \Magento\Customer\Model\Address\Config::getStore
```

```php
public function getStore()
{
    if ($this->_store === null) {
        $this->_store = $this->_storeManager->getStore();
    }
    return $this->_store;
}
```

I guess that's why there was a `$this->storeManager->setCurrentStore($orderStore);` in `\Magento\Sales\Model\Order\Address\Renderer::format`.  

This MR aims to remove the usage of $this->storeManager->setStore which may produce unwanted side-effects and use proper method available in the `\Magento\Customer\Model\Address\Config` API.

### Fixed Issues (if relevant)

Cases where multiple order mails (with different stores) have to be sent by the cron job.
Order addresses are not formatted accordingly to their respective store ID.
Different formats must be set by store ID.

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#38412: Wrong store ID used in Order Address Renderer

### CODE REVIEW

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by 93kamikaze** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
Unit tests needs to be reworked as they don't test expected results but internal implementation 🤡

**Comment by thomas-kl1** [comment]
@magento run all tests

**Review by andrewbess — CHANGES_REQUESTED** [review]
Hello @thomas-kl1 
Thank you for your contribution
Please check [this guide](https://developer.adobe.com/commerce/contributor/guides/code-contributions/). We need to avoid removing|changing protected properties in the @api classes.
This is main reason of the tests failures.
Could you please check failures, documentation and fix your PR?
Thank you in advance.

File: app/code/Magento/Sales/Model/Order/Address/Renderer.php [hunk 1]
```diff
@@ -80,14 +76,13 @@ public function format(Address $address, $type)
...
+        $this->addressConfig->setStore($originalStore);
+
         return $formatType->getRenderer()->renderArray($addressData);
     }
 
     /**
      * Returns locale by storeId
-     *
-     * @param int $storeId
-     * @return string
```
Comment (andrewbess) [inline app/code/Magento/Sales/Model/Order/Address/Renderer.php]: Looks like it is unnecessary removing
Please don't remove PHPDocs like this.

**Review by andrewbess — CHANGES_REQUESTED** [review]
Hello @thomas-kl1 
Thank you for your fast reaction and fixes.
However, let's not touch the class properties and the constructor PHPDocs.
Let's try ti fix only format method.
Also, please check my comment for getLocaleByStoreId method PHPDocs.
Thank you in advance.

**Comment by engcom-Echo** [comment]
@magento create issue

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@thomas-kl1 Can you please have look on build failure of [Integration Tests](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/37932/9142b459e91aeb74533c7fa8a4b19ade/Integration/console-error-logs.html) seems to be caused by PR changes.

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
\Magento\Customer\Model\Metadata\AddressMetadata::getAttributeMetadata
`$attribute = $this->attributeMetadataDataProvider->getAttribute(AddressMetadataInterface::ENTITY_TYPE_ADDRESS, $attributeCode);`

\Magento\Customer\Model\AttributeMetadataDataProvider::getAttribute
`return $this->eavConfig->getAttribute($entityType, $attributeCode);`

\Magento\Eav\Model\Config::initSystemAttributes
```
$attributeCollection = $this->_universalFactory->create($entityType->getEntityAttributeCollection());
$websiteId = $attributeCollection instanceof Collection ? $this->getWebsiteId($attributeCollection) : 0;
```
        

\Magento\Eav\Model\ResourceModel\Attribute\Collection::getWebsite
```
if ($this->_website === null) {
    $this->_website = $this->_storeManager->getStore()->getWebsite();
}
return $this->_website;
```

So we should override the getWebsite method in the following class:
\Magento\Customer\Model\ResourceModel\Address\Attribute\Collection
and use the result of \Magento\Customer\Model\Address\Config::getStore instead

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
I'll squash the commits tomorrow, right now I'm on my phone 😅

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by thomas-kl1** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE

**Review by engcom-Hotel — APPROVED** [review]

**Review by andrewbess — APPROVED** [review]
Hello @thomas-kl1 
Thank you for your contribution.
Fixes look nice for me
Approved

**Comment by andrewbess** [comment]
@magento run Functional Tests B2B

**Comment by andrewbess** [comment]
@magento run Functional Tests B2B

**Comment by andrewbess** [comment]
As for me the failure B2B MFTF tests are not related on these changes.

**Comment by engcom-Echo** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`**

**Manual testing scenario:**
 We have followed steps from [comment](https://github.com/magento/magento2/issues/38412#issuecomment-1931577790)

**Before:** :heavy_multiplication_x: 
Screenshot 2024-03-04 at 5 18 28 PM

**After:** :heavy_check_mark:  
Screenshot 2024-03-04 at 5 20 12 PM

Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Integration Tests,Functional Tests B2B,WebAPI Tests

**Comment by engcom-Echo** [comment]
**Integration Tests** failure are not related to PR changes is known flaky failure.
**Functional Tests B2B** failure are different on last two run on same commit.
Run1
Screenshot 2024-03-05 at 12 45 30 PM

Run2
Screenshot 2024-03-05 at 12 45 34 PM

Hence Moving it to `Merge In Progress`

**Comment by thomas-kl1** [comment]
any update @engcom-Echo when it's going to be merged?

**Comment by engcom-Echo** [comment]
> any update @engcom-Echo when it's going to be merged?

Its already in the process of merging we will update once it it mereged.

**Pull request merged**